sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala

```diff
@@ -191,4 +193,86 @@ class UDFSuite extends QueryTest with SharedSQLContext {
     // pass a decimal to intExpected.
     assert(sql("SELECT intExpected(1.0)").head().getInt(0) === 1)
   }
+
+  private def checkNumUDFs(df: DataFrame, expectedNumUDFs: Int): Unit = {
+    val udfs = df.queryExecution.optimizedPlan.collect {
+      case p: logical.Project => p.projectList.flatMap {
+        case e => e.collect {
+          case udf: ScalaUDF => udf
+        }
+      }
+    }.flatMap(functions => functions)
+    assert(udfs.length === expectedNumUDFs)
+  }
+
+  test("nondeterministic udf: using UDFRegistration") {
+    import org.apache.spark.sql.functions._
+
+    val deterministicUDF = sqlContext.udf.register("plusOne1", (x: Int) => x + 1)
+    val nondeterministicUDF = deterministicUDF.nonDeterministic
+    sqlContext.udf.register("plusOne2", nondeterministicUDF)
+
+    {
+      val df = sql("SELECT 1 as a")
+        .select(col("a"), deterministicUDF(col("a")).as("b"))
+        .select(col("a"), col("b"), deterministicUDF(col("b")).as("c"))
+      checkNumUDFs(df, 3)
```

Comment:
The default value of `foldable` is false, which is why we see three expressions at here.